### PR TITLE
nixos/test-driver: use attrname from nodes to refer to machines

### PR DESCRIPTION
**Pull request by Sohalt**
###### Description of changes

The NixOS test-driver sets up VMs as specified by the `nodes` attrset passed to `makeTest`. The VMs can then be interacted with from the python `testScript` using variables named after `system.name` of the VM. This breaks tests when `system.name` changes (e.g. because you like to add a git SHA to the name).

This change makes it possible to refer to machines by the attrname for the
config in the `nodes` attrset as opposed to the `system.name`, e.g.

```
makeTest = {
    nodes = {
        foo = { pkgs, ... }: {
            ...
        };
        bar = { pkgs, ... }: {
            ...
        };
    };
    testScript = ''
        foo.start()
        bar.start()
        ...
    '';
}
```

I think this way is strictly better, because the attrnames in `nodes` are guaranteed to be unique, they are usually defined right next to the `testScript` and they are completely independent from other parts of the module system (i.e. only the test-driver uses them).

The code is not very clean and not backwards compatible. I just want to have a starting point for discussion.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/testing-python.nix [hunk 1]
```diff
@@ -83,10 +83,9 @@ rec {
...
 
       vlans = map (m: m.config.virtualisation.vlans) (lib.attrValues nodes);
-      vms = map (m: m.config.system.build.vm) (lib.attrValues nodes);
 
       nodeHostNames = let
-        nodesList = map (c: c.config.system.name) (lib.attrValues nodes);
+        nodesList = (lib.attrNames nodes);
       in nodesList ++ lib.optional (lib.length nodesList == 1 && !lib.elem "machine" nodesList) "machine";
 
       # TODO: This is an implementation error and needs fixing
```
Comment (roberth) [inline nixos/lib/testing-python.nix:91]: This comment was once added to acknowledge the problem that these names were too coupled.

When the `nodes.<name>` names correspond to python identifiers only, and to hostnames/system names only by their default value, we can get rid of this comment and `invalidNodeNames`.

We may still want to validate the attribute names to be valid identifiers, but as an alternative, we could use a dict of nodes in the test script. I don't have a much of an opinion about this choice.

File: nixos/lib/testing-python.nix [hunk 2]
```diff
@@ -160,7 +157,9 @@ rec {
         # set defaults through environment
         # see: ./test-driver/test-driver.py argparse implementation
         wrapProgram $out/bin/nixos-test-driver \
-          --set startScripts "''${vmStartScripts[*]}" \
+          --set startScripts "${lib.concatStringsSep "," (lib.mapAttrsToList
+            (name: machine: "${name}=${machine.config.system.build.vm}/bin/run-${machine.config.system.name}-vm")
+            nodes)}" \
```
Comment (roberth) [inline nixos/lib/testing-python.nix:162]: Perhaps passing this as JSON is a little cleaner, especially on the python side.

**Review by roberth — COMMENTED** [review]
I agree about the goal and it doesn't seem to be _terribly_ backwards incompatible.

Note to self: this conflicts superficially with https://github.com/NixOS/nixpkgs/pull/176557; can easily be moved to the new files.

File: nixos/lib/testing-python.nix (hunk 2 above)
Comment (Sohalt) [reply nixos/lib/testing-python.nix:162]: Yes, after I was done with the implementation, I also thought JSON is probably cleaner. I'll try to do another pass on the weekend.

File: nixos/lib/testing-python.nix [hunk 3]
```diff
@@ -83,10 +83,9 @@ rec {
             "nixos-test-driver-${testName}";
 
       vlans = map (m: m.config.virtualisation.vlans) (lib.attrValues nodes);
-      vms = map (m: m.config.system.build.vm) (lib.attrValues nodes);
 
       nodeHostNames = let
-        nodesList = map (c: c.config.system.name) (lib.attrValues nodes);
```
Comment (roberth) [inline nixos/lib/testing-python.nix:89]: This line has moved to `nixos/lib/testing/driver.nix`.

**Pull request closed without merging**